common/dbus/platform/asio_libdbus/dbus_client.cpp

```diff
@@ -0,0 +1,307 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <common/dbus.hpp>
+
+#include <functional>
+#include <memory>
+#include <string>
+
+#include <common/error.hpp>
+#include <common/expected.hpp>
+#include <common/events.hpp>
+#include <common/log.hpp>
+
+#include <boost/asio.hpp>
+#include <dbus/dbus.h>
+
+namespace mender {
+namespace common {
+namespace dbus {
+
+namespace asio = boost::asio;
+
+namespace error = mender::common::error;
+namespace expected = mender::common::expected;
+namespace events = mender::common::events;
+namespace log = mender::common::log;
+
+using namespace std;
+
+// The code below integrates ASIO and libdbus. Or, more precisely, it uses
+// asio::io_context as the main/event loop for libdbus.
+//
+// HandleDispatch() makes sure message dispatch is done. The *Watch() functions
+// allow libdbus to set up and cancel watching of its connection's file
+// descriptor(s). The *Timeout() functions do the same just for
+// timeouts. HandleReply() is a C function we use to extract the DBus reply and
+// pass it to a handler given to DBusClient::CallMethod().
+// (see the individual functions below for details)
+
+// friends can't be static (see class DBusClient)
+void HandleDispatch(DBusConnection *conn, DBusDispatchStatus status, void *data);
+dbus_bool_t AddDBusWatch(DBusWatch *w, void *data);
+static void RemoveDBusWatch(DBusWatch *w, void *data);
+static void ToggleDBusWatch(DBusWatch *w, void *data);
+dbus_bool_t AddDBusTimeout(DBusTimeout *t, void *data);
+static void RemoveDBusTimeout(DBusTimeout *t, void *data);
+static void ToggleDBusTimeout(DBusTimeout *t, void *data);
+static void HandleReply(DBusPendingCall *pending, void *data);
+
+error::Error DBusClient::InitializeConnection() {
+	DBusError dbus_error;
+	dbus_error_init(&dbus_error);
+	dbus_conn_.reset(dbus_bus_get(DBUS_BUS_SYSTEM, &dbus_error));
+	if (!dbus_conn_) {
+		return MakeError(
+			ConnectionError,
+			string("Failed to get connection to system bus: ") + dbus_error.message + "["
+				+ dbus_error.name + "]");
+	}
+
+	dbus_connection_set_exit_on_disconnect(dbus_conn_.get(), FALSE);
+	if (!dbus_connection_set_watch_functions(
+			dbus_conn_.get(), AddDBusWatch, RemoveDBusWatch, ToggleDBusWatch, this, NULL)) {
```

Comment:
Q: Why `NULL` as opposed to `nullptr`? 

I guess it makes not difference really

Reply:
There are implicit conversions from `nullptr` to `NULL`, but I tried to stick with `NULL` when passing the value to a C API.